File: js/src/day1/QuickSort.ts
function qs(arr: number[], low: number, high: number) {
    if (low >= high) {
        return;
    }

    const pivotIdx = partition(arr, low, high);

    qs(arr, low, pivotIdx - 1);
    qs(arr, pivotIdx + 1, high);
}

function partition(arr: number[], low: number, high: number): number {
    const pivot = arr[high];

    let idx = low - 1;

    for (let i = low; i < high; i++) {
        if (arr[i] <= pivot) {
            idx++;
            const tmp = arr[i];
            arr[i] = arr[idx];
            arr[idx] = tmp;
        }
    }

    idx++;
    arr[high] = arr[idx];
    arr[idx] = pivot;
    return idx;
}

export default function quick_sort(arr: number[]): void {
    qs(arr, 0, arr.length - 1);
}

File: golang/sorting/quick_sort.go
package sorting

func qs(arr []int, low, high int) {
	if low >= high {
		return
	}

	pivotIdx := partition(arr, low, high)

	qs(arr, pivotIdx+1, high)
	qs(arr, low, pivotIdx-1)
}

func partition(arr []int, low, high int) int {
	pivot := arr[high]
	idx := low - 1
	for i := low; i < high; i++ {
		if arr[i] <= pivot {
			idx++
			arr[i], arr[idx] = arr[idx], arr[i]
		}
	}

	idx++
	arr[high], arr[idx] = arr[idx], pivot
	return idx
}

func quick_sort(arr []int) []int {
	qs(arr, 0, len(arr)-1)
	return arr
}
